posts/2022-10-19-runtime-multistage-programming.adoc

```diff
@@ -0,0 +1,326 @@
+# Custom JIT compilation with Runtime Multi-Stage Programming
+
+For a while now I've been fiddling with a library called https://github.com/markehammons/slinc[Slinc], a way to link C code in Scala. Since my original version, I've been using context values, macros, and other tools to generate code for writing and reading representations of Structs into and out of native memory.
+
+Originally, I was using Scala 3 macros to produce native bindings and helper code, but I recently discovered that Runtime Multi-Stage Programming produces much better code. My Slinc project therefore is a real-world example of the worth of Runtime Multi-Stage Programming
```

Comment:
> but I recently discovered that Runtime Multi-Stage Programming produces much better code

Maybe you'll explain it further down, but in what way? More performant code? More readable code (even though I'm assuming users aren't reading it), or easier code to use for consumers?

Reply:
The answer is more performant code. Positioning code that had to be queried from an array before (and yes, arrays are already performant) was instead able to be inlined, resulting in a rather gigantic performance boost of around 10x.

Reply:
Explained better upfront